pkg/compose/logs.go

```diff
@@ -20,55 +20,85 @@ import (
 	"context"
 	"io"
 
-	"github.com/docker/compose-cli/pkg/api"
-	"github.com/docker/docker/api/types"
-	"github.com/docker/docker/pkg/stdcopy"
 	"golang.org/x/sync/errgroup"
 
+	"github.com/docker/compose-cli/pkg/api"
 	"github.com/docker/compose-cli/pkg/utils"
+	"github.com/docker/docker/api/types"
+
+	"github.com/docker/docker/pkg/stdcopy"
 )
 
 func (s *composeService) Logs(ctx context.Context, projectName string, consumer api.LogConsumer, options api.LogOptions) error {
 	list, err := s.getContainers(ctx, projectName, oneOffExclude, true, options.Services...)
+	if err != nil {
+		return err
+	}
 
+	project, err := s.projectFromContainerLabels(list, projectName)
 	if err != nil {
 		return err
 	}
+
+	printer := newLogPrinter(consumer)
+
 	eg, ctx := errgroup.WithContext(ctx)
-	for _, c := range list {
-		c := c
-		service := c.Labels[api.ServiceLabel]
-		container, err := s.apiClient.ContainerInspect(ctx, c.ID)
-		if err != nil {
-			return err
-		}
+	eg.Go(func() error {
+		_, err := printer.Run(false, "", nil)
+		return err
+	})
+
+	containers, err := s.getContainers(ctx, projectName, oneOffInclude, false)
+	if err != nil {
+		return err
+	}
 
+	if options.Follow {
 		eg.Go(func() error {
-			r, err := s.apiClient.ContainerLogs(ctx, container.ID, types.ContainerLogsOptions{
-				ShowStdout: true,
-				ShowStderr: true,
-				Follow:     options.Follow,
-				Since:      options.Since,
-				Until:      options.Until,
-				Tail:       options.Tail,
-				Timestamps: options.Timestamps,
+			return s.watchContainers(project, project.ServiceNames(), printer.HandleEvent, containers, func(c types.Container) error {
```

Comment:
project.ServiceNames() -> options.Services

you don't (and we don't want you) need to rebuild project model to collect logs